src/meta/src/hummock/compactor_manager.rs

```diff
@@ -401,79 +302,53 @@ impl CompactorManager {
             .as_secs();
-        let mut guard = self.task_heartbeats.write();
-        if let Some(heartbeats) = guard.get_mut(&context_id) {
-            for progress in progress_list {
-                if let Some(task_ref) = heartbeats.get_mut(&progress.task_id) {
-                    if task_ref.num_ssts_sealed < progress.num_ssts_sealed
-                        || task_ref.num_ssts_uploaded < progress.num_ssts_uploaded
-                        || task_ref.num_progress_key < progress.num_progress_key
-                    {
-                        // Refresh the expiry of the task as it is showing progress.
-                        task_ref.expire_at = now + self.task_expiry_seconds;
-                        task_ref.num_ssts_sealed = progress.num_ssts_sealed;
-                        task_ref.num_ssts_uploaded = progress.num_ssts_uploaded;
-                        task_ref.num_progress_key = progress.num_progress_key;
-                    }
-                    task_ref.num_pending_read_io = progress.num_pending_read_io;
-                    task_ref.num_pending_write_io = progress.num_pending_write_io;
+        let mut cancel_tasks = vec![];
+        for progress in progress_list {
+            if let Some(task_ref) = self.task_heartbeats.get_mut(&progress.task_id) {
+                if task_ref.num_ssts_sealed < progress.num_ssts_sealed
+                    || task_ref.num_ssts_uploaded < progress.num_ssts_uploaded
+                    || task_ref.num_progress_key < progress.num_progress_key
+                {
+                    // Refresh the expiry of the task as it is showing progress.
+                    task_ref.expire_at = now + self.task_expiry_seconds;
+                    task_ref.num_ssts_sealed = progress.num_ssts_sealed;
+                    task_ref.num_ssts_uploaded = progress.num_ssts_uploaded;
+                    task_ref.num_progress_key = progress.num_progress_key;
                 }
-            }
-        }
-    }
-
-    pub fn compactor_num(&self) -> usize {
-        self.policy.read().compactor_num()
-    }
-
-    /// Return the maximum number of tasks that can be assigned to all compactors.
-    pub fn max_concurrent_task_number(&self) -> usize {
-        self.policy.read().max_concurrent_task_num()
-    }
+                task_ref.num_pending_read_io = progress.num_pending_read_io;
+                task_ref.num_pending_write_io = progress.num_pending_write_io;
 
-    /// Update compactor state based on its workload.
-    pub fn update_compactor_state(
-        &self,
-        context_id: HummockContextId,
-        workload: CompactorWorkload,
-    ) {
-        if let Some(compactor) = self.policy.read().get_compactor(context_id) {
-            compactor.cpu_ratio.store(workload.cpu, Ordering::Release);
+                // timeout check
+                if task_ref.expire_at < now {
+                    // cancel
+                    cancel_tasks.push(task_ref.task.clone())
+                }
```

Comment:
This only puts the expired tasks that are included in the `progress_list` in `cancel_tasks`. We miss one case: tasts that are in `self.task_heartbeats` but not in `progress_list` should also be cancelled.

Reply:
Based on the previous discussion, here's the design: the compactor only tries to cancel the tasks that it can sense (in progress_list), while the other tasks that it can't sense (not in progress_list) will be cleaned up by the background threads.